### ISSUE TITLE
Only one thread is used on macOS (super slow on CPU)

### ISSUE THREAD

**Issue by danieleds** [pytorch/pytorch#7686]
Computations on my macbook's CPU are extremely slow.

`torch.get_num_threads()` always says 1, and `torch.set_num_threads(n)` has no effect.

I also tried building from source using the workaround described [here](https://github.com/pytorch/pytorch/issues/6328#issuecomment-383290539) in order to enable libomp support, but nothing changed.

Am I doing something wrong?

**Comment by zou3519** [comment]
cc @cpuhrsch we should enable tbb even when omp is not available (and torch.set_num_threads > 0)

**Comment by cpuhrsch** [comment]
Hello @danieleds,

Could you also add the code you ran to determine that it's slow, or ideally provide a small snippet. Has this set in recently?

Thanks,
Christian

**Comment by danieleds** [comment]
@cpuhrsch I'm just starting to use pytorch so I'm not able to tell if this is a regression.

This is a snippet equivalent to what I'm doing:

```python
for k in range(10):
    x = torch.zeros(50, 6000)
    a = torch.rand(6000, 6000)
    b = torch.rand(6000, 6000)
    c = torch.rand(6000)

    x[-1] = torch.rand(6000)
    for i in range(50-2, -1, -1):
        x[i] = torch.nn.functional.tanh(x[i+1]@a + x[i+1]@b + c)

    print(k)
```

Actually, now that I wrote this demo script it doesn't seem *that* slow, but I have no way to know if the matrix multiplications are being parallelised or not: get_num_threads is still 1.

**Comment by cpuhrsch** [comment]
@danieleds - did you install mkl as part of your source install?

**Comment by danieleds** [comment]
@cpuhrsch Yes

**Comment by cpuhrsch** [comment]
@danieleds You can set the number of threads using OMP_NUM_THREADS as well. You should set that to a value that isn't higher than the number of available threads. You can play around with it to see what works. You can use [perf](https://perf.wiki.kernel.org/index.php/Main_Page) to get an idea of the utilization, e.g. perf stat. [taskset](https://linux.die.net/man/1/taskset) also allows you to set a the number of threads at an OS level.

For example

```
$ numactl --membind=0 taskset -c 0-9 perf stat python /tmp/asd
num_threads: 10
0
1
2
3
4
5
6
7
8
9

 Performance counter stats for 'python /tmp/asd':

     102703.859727      task-clock (msec)         #    6.493 CPUs utilized
         6,648,917      context-switches          #    0.065 M/sec
            25,449      cpu-migrations            #    0.248 K/sec
            23,129      page-faults               #    0.225 K/sec
   225,915,753,932      cycles                    #    2.200 GHz
   <not supported>      stalled-cycles-frontend
   <not supported>      stalled-cycles-backend
   217,365,021,721      instructions              #    0.96  insns per cycle
    52,707,870,505      branches                  #  513.202 M/sec
        33,402,160      branch-misses             #    0.06% of all branches

      15.818744212 seconds time elapsed
```

**Comment by danieleds** [comment]
Thanks. OMP_NUM_THREADS seems to do the trick (even though get_num_threads is still 1).

Performance actually gets worse (even with OMP_NUM_THREADS=2... and I clearly have 2 free processing units), but at least the CPU usage now rises above 100%. I assume that's because my matrices are too small.

**Comment by cpuhrsch** [comment]
@danieleds How many cores and threads does your computer have?

**Comment by danieleds** [comment]
@cpuhrsch 2 cores, 4 threads

**Comment by Pawel-Marcin-Chojnacki** [comment]
I used 
$ export NUM_CORES=2
$ export MKL_NUM_THREADS=$NUM_CORES OMP_NUM_THREADS=$NUM_CORES
and it started to taking all advantage of my Macbook Air. (CPU util 200%).

**Comment by wangxiaoying** [comment]
Got the same problem here. I compiled pytorch with openblas on mac with cpu and cannot change the number of thread in both python interface and c++ interface. OMP_NUM_THREADS does not help.

**Comment by benkent100** [comment]
I'm still finding this with the stable 1.5 macos libtorch download.

torch::get_num_threads() added to the basic example in the tutorial on the c++ front end (https://pytorch.org/tutorials/advanced/cpp_frontend.html) always returns 1 and a more compute intensive example doesn't use more than 100% CPU.

I've tried calling init_num_threads() before this, and setting the OMP/MKL_NUM_THREADS environment variables, but nothing seems to work.

Is there another fix?

**Comment by cpuhrsch** [comment]
cc @VitalyFedyunin

**Comment by nikotsingos** [comment]
Hi there, just curious, this is still an issue with the stable 1.6.0 macos release, right ?
with the same code, after torch::init_num_threads(),  i am seeing intraop threads= 4 and  interop threads =2  on a docker ubuntu container (docker being setup for 4 cpus)  but on my native mac (8 cores mbpro 10.15.3) , libtorch appears to go with intraop threads= 1 and interop threads = 8.  The net result being that it runs much slower native than the docker container version...

**Comment by lunixbochs** [comment]
Torch 1.6.0 from `pip` is still single threaded on Mac out of the box for me :(

**Comment by bangbangjim** [comment]
any solution to this problem for Mac yet?

**Comment by benkent100** [comment]
We ended up rebuilding the libraries from source and forcing it to use openmp.

**Comment by lunixbochs** [comment]
did you do anything specific to force it to use openmp? would love even some vague instructions, as I'm interested in doing that too

**Comment by rfeinman** [comment]
@jimmyjimmy94 @benkent100 @lunixbochs I've collected some insight on this after struggling for a few days to get a working build on macOS with MKL + OpenMP threading.

The problem I've identified is that the OpenMP header files are not available on macOS by default (they do not come with the Apple Clang compiler). When you install the [pytorch dependencies via conda](https://github.com/pytorch/pytorch#install-dependencies), you'll get intel's openmp runtime library (libiomp5.dylib), but you will NOT get the associated header files. AFAIK, intel's openmp header files are not available in any conda package. As such, [FindOpenMP.cmake](https://github.com/pytorch/pytorch/blob/master/cmake/Modules/FindOpenMP.cmake) will fail to find openmp during the build.

There's a workaround described for Windows in this [PyTorch tutorial](https://pytorch.org/docs/stable/notes/windows.html?highlight=windows). PyTorch team seems to have created a zip folder for Windows users (called mkl_2020.0.166.7z) that includes the necessary openmp header files. When downloaded and added to CMAKE_INCLUDE_PATH, this fixes the problem during the cmake build. Annoyingly, there is no equivalent tutorial for macOS users, and mac users may not think to look at this page. 

I had success by using the same openmp header files from this zip folder. You can either add them to your conda environment's `include/` directory (maybe not the best idea, since it could interfere with conda package management) or simply place them in some other directory which you include in your CMAKE_INCLUDE_PATH.

If the header files from this zip folder don't work, you can also download them directly by installing the Intel oneAPI HPC toolkit. I hope this is helpful.

**Comment by yayuntian** [comment]
I have worked hard to check the problem, comparing the performance of linux

**Comment by malfet** [comment]
macOS CPU is no longer limited to a single thread — `torch.get_num_threads()` returns the core count and OpenMP is enabled on the released macOS build:

```
$ python -c "import torch; print(torch.__version__, torch.backends.openmp.is_available(), torch.get_num_threads())"
2.13.0 True 6
```

Closing as long since implemented.
